### DISCUSSION TITLE
v8 and event loop

### DISCUSSION THREAD

**Discussion by nextfool** [ClearFoundry/ClearScript#323] [General]
Hi @ClearScriptLib ,

Since V8 does not support any web api(setTimeout, postMessage etc.), I have to implement them myself. So I have created my own "event loop".  But V8 DOES support promises, and thus it has its own "microtask queue", and it is out of my control.

Now my questions are:

1.How do I access v8's  "microtask queue" and integrate it into my "event loop"? I guess it is not possible isn't?
2. Does Clearscript have any infrastructure to allow me to extend v8's internal queue?
3. Do you have plan to include "libuv"(used by Nodejs) or any other similar component into Clearscript, so that it has the "official" event loop?

So basically what I need is a way to "sync" between the promises and my "event loop", so that the script runs in my engine would feel like it's running in Nodejs or a browser.

Thank you.

**Comment by ClearScriptLib** [comment]
Hi @nextfool,

>How do I access v8's "microtask queue" and integrate it into my "event loop"?
>Does Clearscript have any infrastructure to allow me to extend v8's internal queue?

ClearScript doesn't expose script engine internals.

>Do you have plan to include "libuv"(used by Nodejs) or any other similar component into Clearscript, so that it has the "official" event loop?

The .NET platform includes asynchronous I/O and synchronization facilities as part of its standard API set.

Our [worker sample](https://github.com/microsoft/ClearScriptWorkerSample) demonstrates how asynchronous event loops and cross-runtime communications can be implemented without engine internals or additional libraries. Please let us know what it's missing.

>So basically what I need is a way to "sync" between the promises and my "event loop", so that the script runs in my engine would feel like it's running in Nodejs or a browser.

Please clarify. What would you like to do, specifically, that the existing API doesn't allow you to do?

Thanks!

**Reply by nextfool** [reply]
hi @ClearScriptLib ,
sorry for the late reply, as I was digging into the node source and tried to learn how node implement event loop.  As now I feel like I asked the wrong question here, it should be:
How do I handle the v8 micro task queue with clear script?

Given the v8 document about " [MicrotasksPolicy](https://v8docs.nodesource.com/node-14.1/d2/dc3/namespacev8.html#af774e70316f1ab21a93eaab1c57316a8)" here, I know that for a embedder of v8, if you want to handle the microtask queue yourself,  you need to use the "MicrotasksPolicy::kExplici", instead of the default value "MicrotasksPolicy::kAuto ". (please correct me if wrong) 
Please see following the nodejs source code:
![image](https://user-images.githubusercontent.com/2167369/142979798-bddb6b91-a532-43c9-8f4d-500c3d279d3c.png)

I searched the terms like "MicrotasksPolicy" in clearscript source, and can not find any thing. So I presume that clearscript just uses the default value i.e.   "MicrotasksPolicy::kAuto "

In this case, if I want to implement a "node like" event loop (timer callback, io events, imediates, close handlers), how do I invoke the "microtask queue" in between each "phases"?

Thank you.

**Reply by nextfool** [reply]
another issue with current case would be, I can not implement the ["queueMicrotask"](https://developer.mozilla.org/en-US/docs/Web/API/queueMicrotask)

**Comment by ClearScriptLib** [comment]
Hi @nextfool,

>how do I invoke the "microtask queue" in between each "phases"?

You don't have to. V8 processes the microtask queue automatically when it exits from the outermost script scope.

>another issue with current case would be, I can not implement the "queueMicrotask"

You can use promises to implement such a function:

``` C#
dynamic addQueueMicrotask = engine.Evaluate(@"(function (onError) {
    queueMicrotask = (function (fn) {
        Promise.resolve().then(fn).catch(onError);
        return undefined;
    }).bind();
})");
addQueueMicrotask(new Action<object>(error => {
    var message = error is ScriptObject obj ? obj.InvokeMethod("toString") : error.ToString();
    Console.WriteLine("Uncaught " + message);
}));
```

Please let us know if we're missing something.

Thanks!
